File: packages/cli/test/services/repo_artifacts_test.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:dotweave/src/config/constants.dart';
import 'package:dotweave/src/config/platform.dart';
import 'package:dotweave/src/config/sync_schema.dart';
import 'package:dotweave/src/lib/error.dart';
import 'package:dotweave/src/lib/posix_chmod.dart';
import 'package:dotweave/src/services/repo_artifacts.dart';
import 'package:dotweave/src/services/sync_context.dart';
import 'package:path/path.dart' as p;
import 'package:test/test.dart';

void main() {
  final temporaryDirectories = <String>[];

  Future<String> createWorkspace() async {
    final directory = await Directory.systemTemp.createTemp(
      'dotweave-repo-artifacts-',
    );

    temporaryDirectories.add(directory.path);

    return directory.path;
  }

  ResolvedSyncConfigEntry createConfigEntry({
    PlatformStringValue configuredLocalPath = const PlatformStringValue(
      defaultValue: '~/.config/app',
    ),
    PlatformSyncMode configuredMode = const PlatformSyncMode(
      defaultValue: 'normal',
    ),
    ConfiguredSyncRepoPath? configuredRepoPath,
    SyncConfigEntryKind kind = 'directory',
    String localPath = '/home/user/.config/app',
    SyncMode mode = 'normal',
    List<String> profiles = const [],
    bool profilesExplicit = false,
    String repoPath = '.config/app',
  }) {
    return ResolvedSyncConfigEntry(
      configuredLocalPath: configuredLocalPath,
      configuredMode: configuredMode,
      configuredRepoPath: configuredRepoPath,
      kind: kind,
      localPath: localPath,
      mode: mode,
      modeExplicit: true,
      permissionExplicit: false,
      profiles: profiles,
      profilesExplicit: profilesExplicit,
      repoPath: repoPath,
    );
  }

  EffectiveSyncConfig createConfig(
    ResolvedSyncConfigEntry entry, [
    String? activeProfile,
  ]) {
    return EffectiveSyncConfig(
      activeProfile: activeProfile,
      age: const RuntimeAgeConfig(identityFile: 'keys.txt', recipients: []),
      entries: [entry],
      profiles: const [],
      version: AppConstants.sync.configVersion,
    );
  }

  ArtifactOwnershipConfig ownershipOf(EffectiveSyncConfig config) {
    return (entries: config.entries, profiles: config.profiles);
  }

  tearDown(() async {
    while (temporaryDirectories.isNotEmpty) {
      final directory = temporaryDirectories.removeLast();

      if (await Directory(directory).exists()) {
        await Directory(directory).delete(recursive: true);
      }
    }
  });

  group('repo-artifacts service', () {
    test('collects artifact profiles from entries', () {
      final entries = [
        createConfigEntry(profiles: ['profile-a', 'profile-b']),
        createConfigEntry(profiles: ['profile-b', 'profile-c']),
      ];
      final profiles = collectArtifactProfiles(entries);
      expect(profiles.contains(AppConstants.sync.defaultProfile), isTrue);
      expect(profiles.contains('profile-a'), isTrue);
      expect(profiles.contains('profile-b'), isTrue);
      expect(profiles.contains('profile-c'), isTrue);
      expect(profiles.length, 4);
    });

    test('builds artifact keys correctly', () {
      expect(
        buildArtifactKey(
          const DirectoryRepoArtifact(profile: 'default', repoPath: 'config'),
        ),
        'default/config/',
      );

      expect(
        buildArtifactKey(
          FileRepoArtifact(
            category: 'plain',
            contents: Uint8List(0),
            executable: false,
            profile: 'default',
            repoPath: 'file.txt',
          ),
        ),
        'default/file.txt',
      );
    });

    test(
      'keeps artifact keys logical while resolving physical profiles paths',
      () {
        expect(
          buildArtifactKey(
            FileRepoArtifact(
              category: 'plain',
              contents: Uint8List(0),
              executable: false,
              profile: 'work',
              repoPath: '.gitconfig',
            ),
          ),
          'work/.gitconfig',
        );

        expect(
          buildArtifactKey(
            const DirectoryRepoArtifact(
              profile: 'work',
              repoPath: '.config/app',
            ),
          ),
          'work/.config/app/',
        );

        expect(
          buildArtifactKey(
            FileRepoArtifact(
              category: 'secret',
              contents: Uint8List(0),
              executable: false,
              profile: 'work',
              repoPath: '.ssh/id',
            ),
          ),
          'work/.ssh/id${AppConstants.sync.secretArtifactSuffix}',
        );

        expect(
          resolveArtifactRelativePath(
            category: 'plain',
            profile: 'work',
            repoPath: '.gitconfig',
          ),
          'profiles/work/.gitconfig',
        );

        expect(
          resolveArtifactRelativePath(
            category: 'secret',
            profile: 'work',
            repoPath: '.ssh/id',
          ),
          'profiles/work/.ssh/id${AppConstants.sync.secretArtifactSuffix}',
        );
      },
    );

    test('identifies secret artifact paths', () {
      expect(
        isSecretArtifactPath(
          'file.txt${AppConstants.sync.secretArtifactSuffix}',
        ),
        isTrue,
      );
      expect(isSecretArtifactPath('file.txt'), isFalse);
    });

    test('strips secret artifact suffix', () {
      final path = 'file.txt${AppConstants.sync.secretArtifactSuffix}';
      expect(stripSecretArtifactSuffix(path), 'file.txt');
      expect(stripSecretArtifactSuffix('file.txt'), isNull);
    });

    test('resolves artifact relative paths', () {
      expect(
        resolveArtifactRelativePath(
          category: 'plain',
          profile: 'work',
          repoPath: 'config',
        ),
        'profiles/work/config',
      );

      expect(
        resolveArtifactRelativePath(
          category: 'secret',
          profile: 'work',
          repoPath: 'secrets.json',
        ),
        'profiles/work/secrets.json${AppConstants.sync.secretArtifactSuffix}',
      );
    });

    test('parses artifact relative paths', () {
      const relativePath = 'profiles/home/.bashrc';
      final parsed = parseArtifactRelativePath(relativePath);
      expect(parsed.profile, 'home');
      expect(parsed.repoPath, '.bashrc');
      expect(parsed.secret, isFalse);

      final secretPath =
          'profiles/work/token${AppConstants.sync.secretArtifactSuffix}';
      final parsedSecret = parseArtifactRelativePath(secretPath);
      expect(parsedSecret.profile, 'work');
      expect(parsedSecret.repoPath, 'token');
      expect(parsedSecret.secret, isTrue);
    });

    test('parses only physical profiles artifact paths', () {
      expect(parseArtifactRelativePath('profiles/work/.gitconfig'), (
        profile: 'work',
        repoPath: '.gitconfig',
        secret: false,
        symlink: false,
      ));

      expect(
        parseArtifactRelativePath(
          'profiles/work/.ssh/id${AppConstants.sync.secretArtifactSuffix}',
        ),
        (profile: 'work', repoPath: '.ssh/id', secret: true, symlink: false),
      );

      expect(
        parseArtifactRelativePath(
          'profiles/work/.claude/skills'
          '${AppConstants.sync.symlinkArtifactSuffix}',
        ),
        (
          profile: 'work',
          repoPath: '.claude/skills',
          secret: false,
          symlink: true,
        ),
      );

      expect(
        () => parseArtifactRelativePath('docs/readme.md'),
        throwsA(isA<DotweaveError>()),
      );
      expect(
        () => parseArtifactRelativePath('work/.gitconfig'),
        throwsA(isA<DotweaveError>()),
      );
    });

    test('classifies all-platform ignored artifacts as prunable', () {
      final entry = createConfigEntry(
        configuredMode: const PlatformSyncMode(defaultValue: 'ignore'),
        mode: 'ignore',
        repoPath: '.config/app/settings.json',
        kind: 'file',
      );
      final config = createConfig(entry);

      expect(
        classifyArtifactOwnership(
          config,
          ownershipOf(config),
          parseArtifactRelativePath(
            'profiles/default/.config/app/settings.json',
          ),
          'file',
        ),
        'ignored-prunable',
      );
    });

    test('classifies platform-ignored artifacts owned by another platform as '
        'protected', () {
      final entry = createConfigEntry(
        configuredMode: const PlatformSyncMode(
          defaultValue: 'normal',
          win: 'ignore',
        ),
        configuredRepoPath: const PlatformStringValue(
          defaultValue: '.config/app',
          win: 'AppData/Roaming/app',
        ),
        mode: 'ignore',
        repoPath: 'AppData/Roaming/app',
      );
      final config = createConfig(entry);

      expect(
        classifyArtifactOwnership(
          config,
          ownershipOf(config),
          parseArtifactRelativePath(
            'profiles/default/.config/app/settings.json',
          ),
          'file',
        ),
        'platform-protected',
      );
    });

    test('classifies platform-specific ignored artifacts without another owner '
        'as prunable', () {
      final entry = createConfigEntry(
        configuredMode: const PlatformSyncMode(
          defaultValue: 'normal',
          win: 'ignore',
        ),
        configuredRepoPath: const PlatformStringValue(
          defaultValue: '.config/app',
          win: 'AppData/Roaming/app',
        ),
        mode: 'ignore',
        repoPath: 'AppData/Roaming/app',
      );
      final config = createConfig(entry);

      expect(
        classifyArtifactOwnership(
          config,
          ownershipOf(config),
          parseArtifactRelativePath(
            'profiles/default/AppData/Roaming/app/settings.json',
          ),
          'file',
        ),
        'ignored-prunable',
      );
    });

    test(
      'classifies WSL-ignored linux/default artifacts as platform-protected',
      () {
        final entry = createConfigEntry(
          configuredMode: const PlatformSyncMode(
            defaultValue: 'normal',
            linux: 'secret',
            wsl: 'ignore',
          ),
          configuredRepoPath: const PlatformStringValue(
            defaultValue: '.config/app',
            linux: '.config/app',
            wsl: 'Ubuntu/app',
          ),
          mode: 'ignore',
          repoPath: 'Ubuntu/app',
        );
        final config = createConfig(entry);

        expect(
          classifyArtifactOwnership(
            config,
            ownershipOf(config),
            parseArtifactRelativePath(
              'profiles/default/.config/app/settings.json',
            ),
            'file',
          ),
          'platform-protected',
        );
      },
    );

    test(
      'classifies inactive-profile artifacts with an entry as protected',
      () {
        final entry = createConfigEntry(
          configuredMode: const PlatformSyncMode(defaultValue: 'ignore'),
          kind: 'file',
          mode: 'ignore',
          profiles: ['work'],
          profilesExplicit: true,
          repoPath: '.gitconfig',
        );
        final config = createConfig(entry);

        expect(
          classifyArtifactOwnership(
            config,
            ownershipOf(config),
            parseArtifactRelativePath('profiles/work/.gitconfig'),
            'file',
          ),
          'platform-protected',
        );
      },
    );

    test('collects physical profiles artifacts as logical keys', () async {
      final workspace = await createWorkspace();
      final config = EffectiveSyncConfig(
        activeProfile: AppConstants.sync.defaultProfile,
        age: const RuntimeAgeConfig(identityFile: 'keys.txt', recipients: []),
        entries: const [],
        profiles: const [],
        version: AppConstants.sync.configVersion,
      );

      await Directory(
        p.join(workspace, 'profiles', 'work', '.config', 'app'),
      ).create(recursive: true);
      await File(
        p.join(workspace, 'profiles', 'work', '.gitconfig'),
      ).writeAsString('data');
      await File(
        p.join(
          workspace,
          'profiles',
          'work',
          '.config',
          'app',
          'settings.json',
        ),
      ).writeAsString('{}\n');
      await Directory(p.join(workspace, 'docs')).create(recursive: true);
      await File(
        p.join(workspace, 'docs', 'readme.md'),
      ).writeAsString('support docs\n');
      await Directory(
        p.join(workspace, 'profiles', '.github'),
      ).create(recursive: true);
      await File(
        p.join(workspace, 'profiles', '.github', 'workflow.yml'),
      ).writeAsString('name\n');

      expect(await collectExistingArtifactKeys(workspace, config), {
        'work/.config/app/settings.json',
        'work/.gitconfig',
      });
    });

    test(
      'treats non-executable artifact permission noise as current',
      () async {
        if (Platform.isWindows) {
          return;
        }

        final workspace = await createWorkspace();
        final artifactDirectory = p.join(workspace, 'profiles', 'default');
        final artifactPath = p.join(artifactDirectory, 'file.txt');

        await Directory(artifactDirectory).create(recursive: true);
        await File(artifactPath).writeAsString('data\n');
        posixChmod(artifactPath, 0x180); // 0o600

        expect(
          await isRepoArtifactCurrent(
            workspace,
            FileRepoArtifact(
              category: 'plain',
              contents: Uint8List.fromList(utf8.encode('data\n')),
              executable: false,
              profile: 'default',
              repoPath: 'file.txt',
            ),
          ),
          isTrue,
        );
      },
    );

    test(
      'treats executable artifacts as current when the executable bit matches',
      () async {
        if (Platform.isWindows) {
          return;
        }

        final workspace = await createWorkspace();
        final artifactDirectory = p.join(workspace, 'profiles', 'default');
        final artifactPath = p.join(artifactDirectory, 'tool');

        await Directory(artifactDirectory).create(recursive: true);
        await File(artifactPath).writeAsString('#!/bin/sh\n');
        posixChmod(artifactPath, 0x1ED); // 0o755

        expect(
          await isRepoArtifactCurrent(
            workspace,
            FileRepoArtifact(
              category: 'plain',
              contents: Uint8List.fromList(utf8.encode('#!/bin/sh\n')),
              executable: true,
              profile: 'default',
              repoPath: 'tool',
            ),
          ),
          isTrue,
        );
      },
    );

    test(
      'reports executable artifact drift when the executable bit differs',
      () async {
        if (Platform.isWindows) {
          return;
        }

        final workspace = await createWorkspace();
        final artifactDirectory = p.join(workspace, 'profiles', 'default');
        final artifactPath = p.join(artifactDirectory, 'tool');

        await Directory(artifactDirectory).create(recursive: true);
        await File(artifactPath).writeAsString('#!/bin/sh\n');
        posixChmod(artifactPath, 0x1A4); // 0o644

        expect(
          await isRepoArtifactCurrent(
            workspace,
            FileRepoArtifact(
              category: 'plain',
              contents: Uint8List.fromList(utf8.encode('#!/bin/sh\n')),
              executable: true,
              profile: 'default',
              repoPath: 'tool',
            ),
          ),
          isFalse,
        );
      },
    );
  });
}

File: packages/cli/src/services/repo-artifacts.test.ts
import { chmod, mkdir, rm, writeFile } from "node:fs/promises";
import { join } from "node:path";
import { afterEach, describe, expect, it } from "vitest";
import { AppConstants } from "#app/config/constants.ts";
import type { ResolvedSyncConfigEntry } from "#app/config/sync-schema.ts";
import { createTemporaryDirectory } from "#app/test/helpers/sync-fixture.ts";
import {
  buildArtifactKey,
  classifyArtifactOwnership,
  collectArtifactProfiles,
  collectExistingArtifactKeys,
  isRepoArtifactCurrent,
  isSecretArtifactPath,
  parseArtifactRelativePath,
  resolveArtifactRelativePath,
  stripSecretArtifactSuffix,
} from "./repo-artifacts.ts";
import type { EffectiveSyncConfig } from "./sync-context.ts";

const temporaryDirectories: string[] = [];

const createWorkspace = async () => {
  const directory = await createTemporaryDirectory("dotweave-repo-artifacts-");

  temporaryDirectories.push(directory);

  return directory;
};

const createConfigEntry = (
  overrides: Partial<ResolvedSyncConfigEntry> = {},
): ResolvedSyncConfigEntry => {
  return {
    configuredLocalPath: { default: "~/.config/app" },
    configuredMode: { default: "normal" },
    kind: "directory",
    localPath: "/home/user/.config/app",
    mode: "normal",
    modeExplicit: true,
    permissionExplicit: false,
    profiles: [],
    profilesExplicit: false,
    repoPath: ".config/app",
    ...overrides,
  };
};

const createConfig = (
  entry: ResolvedSyncConfigEntry,
  activeProfile?: string,
): EffectiveSyncConfig => {
  return {
    ...(activeProfile === undefined ? {} : { activeProfile }),
    age: { identityFile: "keys.txt", recipients: [] },
    entries: [entry],
    profiles: [],
    version: AppConstants.SYNC.CONFIG_VERSION,
  };
};

afterEach(async () => {
  while (temporaryDirectories.length > 0) {
    const directory = temporaryDirectories.pop();

    if (directory !== undefined) {
      await rm(directory, { force: true, recursive: true });
    }
  }
});

describe("repo-artifacts service", () => {
  it("collects artifact profiles from entries", () => {
    const entries: Pick<ResolvedSyncConfigEntry, "profiles">[] = [
      { profiles: ["profile-a", "profile-b"] },
      { profiles: ["profile-b", "profile-c"] },
    ];
    const profiles = collectArtifactProfiles(entries);
    expect(profiles.has(AppConstants.SYNC.DEFAULT_PROFILE)).toBe(true);
    expect(profiles.has("profile-a")).toBe(true);
    expect(profiles.has("profile-b")).toBe(true);
    expect(profiles.has("profile-c")).toBe(true);
    expect(profiles.size).toBe(4);
  });

  it("builds artifact keys correctly", () => {
    expect(
      buildArtifactKey({
        kind: "directory",
        profile: "default",
        repoPath: "config",
        category: "plain",
      }),
    ).toBe("default/config/");

    expect(
      buildArtifactKey({
        kind: "file",
        profile: "default",
        repoPath: "file.txt",
        category: "plain",
        contents: new Uint8Array(),
        executable: false,
      }),
    ).toBe("default/file.txt");
  });

  it("keeps artifact keys logical while resolving physical profiles paths", () => {
    expect(
      buildArtifactKey({
        category: "plain",
        contents: new Uint8Array(),
        executable: false,
        kind: "file",
        profile: "work",
        repoPath: ".gitconfig",
      }),
    ).toBe("work/.gitconfig");

    expect(
      buildArtifactKey({
        category: "plain",
        kind: "directory",
        profile: "work",
        repoPath: ".config/app",
      }),
    ).toBe("work/.config/app/");

    expect(
      buildArtifactKey({
        category: "secret",
        contents: new Uint8Array(),
        executable: false,
        kind: "file",
        profile: "work",
        repoPath: ".ssh/id",
      }),
    ).toBe(`work/.ssh/id${AppConstants.SYNC.SECRET_ARTIFACT_SUFFIX}`);

    expect(
      resolveArtifactRelativePath({
        category: "plain",
        profile: "work",
        repoPath: ".gitconfig",
      }),
    ).toBe("profiles/work/.gitconfig");

    expect(
      resolveArtifactRelativePath({
        category: "secret",
        profile: "work",
        repoPath: ".ssh/id",
      }),
    ).toBe(`profiles/work/.ssh/id${AppConstants.SYNC.SECRET_ARTIFACT_SUFFIX}`);
  });

  it("identifies secret artifact paths", () => {
    expect(
      isSecretArtifactPath(
        `file.txt${AppConstants.SYNC.SECRET_ARTIFACT_SUFFIX}`,
      ),
    ).toBe(true);
    expect(isSecretArtifactPath("file.txt")).toBe(false);
  });

  it("strips secret artifact suffix", () => {
    const path = `file.txt${AppConstants.SYNC.SECRET_ARTIFACT_SUFFIX}`;
    expect(stripSecretArtifactSuffix(path)).toBe("file.txt");
    expect(stripSecretArtifactSuffix("file.txt")).toBeUndefined();
  });

  it("resolves artifact relative paths", () => {
    expect(
      resolveArtifactRelativePath({
        category: "plain",
        profile: "work",
        repoPath: "config",
      }),
    ).toBe("profiles/work/config");

    expect(
      resolveArtifactRelativePath({
        category: "secret",
        profile: "work",
        repoPath: "secrets.json",
      }),
    ).toBe(
      `profiles/work/secrets.json${AppConstants.SYNC.SECRET_ARTIFACT_SUFFIX}`,
    );
  });

  it("parses artifact relative paths", () => {
    const relativePath = "profiles/home/.bashrc";
    const parsed = parseArtifactRelativePath(relativePath);
    expect(parsed.profile).toBe("home");
    expect(parsed.repoPath).toBe(".bashrc");
    expect(parsed.secret).toBe(false);

    const secretPath = `profiles/work/token${AppConstants.SYNC.SECRET_ARTIFACT_SUFFIX}`;
    const parsedSecret = parseArtifactRelativePath(secretPath);
    expect(parsedSecret.profile).toBe("work");
    expect(parsedSecret.repoPath).toBe("token");
    expect(parsedSecret.secret).toBe(true);
  });

  it("parses only physical profiles artifact paths", () => {
    expect(parseArtifactRelativePath("profiles/work/.gitconfig")).toEqual({
      profile: "work",
      repoPath: ".gitconfig",
      secret: false,
      symlink: false,
    });

    expect(
      parseArtifactRelativePath(
        `profiles/work/.ssh/id${AppConstants.SYNC.SECRET_ARTIFACT_SUFFIX}`,
      ),
    ).toEqual({
      profile: "work",
      repoPath: ".ssh/id",
      secret: true,
      symlink: false,
    });

    expect(
      parseArtifactRelativePath(
        `profiles/work/.claude/skills${AppConstants.SYNC.SYMLINK_ARTIFACT_SUFFIX}`,
      ),
    ).toEqual({
      profile: "work",
      repoPath: ".claude/skills",
      secret: false,
      symlink: true,
    });

    expect(() => parseArtifactRelativePath("docs/readme.md")).toThrow();
    expect(() => parseArtifactRelativePath("work/.gitconfig")).toThrow();
  });

  it("classifies all-platform ignored artifacts as prunable", () => {
    const entry = createConfigEntry({
      configuredMode: { default: "ignore" },
      mode: "ignore",
      repoPath: ".config/app/settings.json",
      kind: "file",
    });
    const config = createConfig(entry);

    expect(
      classifyArtifactOwnership(
        config,
        config,
        parseArtifactRelativePath("profiles/default/.config/app/settings.json"),
        "file",
      ),
    ).toBe("ignored-prunable");
  });

  it("classifies platform-ignored artifacts owned by another platform as protected", () => {
    const entry = createConfigEntry({
      configuredMode: { default: "normal", win: "ignore" },
      configuredRepoPath: {
        default: ".config/app",
        win: "AppData/Roaming/app",
      },
      mode: "ignore",
      repoPath: "AppData/Roaming/app",
    });
    const config = createConfig(entry);

    expect(
      classifyArtifactOwnership(
        config,
        config,
        parseArtifactRelativePath("profiles/default/.config/app/settings.json"),
        "file",
      ),
    ).toBe("platform-protected");
  });

  it("classifies platform-specific ignored artifacts without another owner as prunable", () => {
    const entry = createConfigEntry({
      configuredMode: { default: "normal", win: "ignore" },
      configuredRepoPath: {
        default: ".config/app",
        win: "AppData/Roaming/app",
      },
      mode: "ignore",
      repoPath: "AppData/Roaming/app",
    });
    const config = createConfig(entry);

    expect(
      classifyArtifactOwnership(
        config,
        config,
        parseArtifactRelativePath(
          "profiles/default/AppData/Roaming/app/settings.json",
        ),
        "file",
      ),
    ).toBe("ignored-prunable");
  });

  it("classifies WSL-ignored linux/default artifacts as platform-protected", () => {
    const entry = createConfigEntry({
      configuredMode: { default: "normal", linux: "secret", wsl: "ignore" },
      configuredRepoPath: {
        default: ".config/app",
        linux: ".config/app",
        wsl: "Ubuntu/app",
      },
      mode: "ignore",
      repoPath: "Ubuntu/app",
    });
    const config = createConfig(entry);

    expect(
      classifyArtifactOwnership(
        config,
        config,
        parseArtifactRelativePath("profiles/default/.config/app/settings.json"),
        "file",
      ),
    ).toBe("platform-protected");
  });

  it("classifies inactive-profile artifacts with an entry as protected", () => {
    const entry = createConfigEntry({
      configuredMode: { default: "ignore" },
      kind: "file",
      mode: "ignore",
      profiles: ["work"],
      profilesExplicit: true,
      repoPath: ".gitconfig",
    });
    const config = createConfig(entry);

    expect(
      classifyArtifactOwnership(
        config,
        config,
        parseArtifactRelativePath("profiles/work/.gitconfig"),
        "file",
      ),
    ).toBe("platform-protected");
  });

  it("collects physical profiles artifacts as logical keys", async () => {
    const workspace = await createWorkspace();
    const config: EffectiveSyncConfig = {
      activeProfile: AppConstants.SYNC.DEFAULT_PROFILE,
      age: { identityFile: "keys.txt", recipients: [] },
      entries: [],
      profiles: [],
      version: AppConstants.SYNC.CONFIG_VERSION,
    };

    await mkdir(join(workspace, "profiles", "work", ".config", "app"), {
      recursive: true,
    });
    await writeFile(join(workspace, "profiles", "work", ".gitconfig"), "data");
    await writeFile(
      join(workspace, "profiles", "work", ".config", "app", "settings.json"),
      "{}\n",
    );
    await mkdir(join(workspace, "docs"), { recursive: true });
    await writeFile(join(workspace, "docs", "readme.md"), "support docs\n");
    await mkdir(join(workspace, "profiles", ".github"), { recursive: true });
    await writeFile(
      join(workspace, "profiles", ".github", "workflow.yml"),
      "name\n",
    );

    await expect(
      collectExistingArtifactKeys(workspace, config),
    ).resolves.toEqual(
      new Set(["work/.config/app/settings.json", "work/.gitconfig"]),
    );
  });

  it.skipIf(process.platform === "win32")(
    "treats non-executable artifact permission noise as current",
    async () => {
      const workspace = await createWorkspace();
      const artifactDirectory = join(workspace, "profiles", "default");
      const artifactPath = join(artifactDirectory, "file.txt");

      await mkdir(artifactDirectory, { recursive: true });
      await writeFile(artifactPath, "data\n", "utf8");
      await chmod(artifactPath, 0o600);

      await expect(
        isRepoArtifactCurrent(workspace, {
          category: "plain",
          contents: Buffer.from("data\n"),
          executable: false,
          kind: "file",
          profile: "default",
          repoPath: "file.txt",
        }),
      ).resolves.toBe(true);
    },
  );

  it.skipIf(process.platform === "win32")(
    "treats executable artifacts as current when the executable bit matches",
    async () => {
      const workspace = await createWorkspace();
      const artifactDirectory = join(workspace, "profiles", "default");
      const artifactPath = join(artifactDirectory, "tool");

      await mkdir(artifactDirectory, { recursive: true });
      await writeFile(artifactPath, "#!/bin/sh\n", "utf8");
      await chmod(artifactPath, 0o755);

      await expect(
        isRepoArtifactCurrent(workspace, {
          category: "plain",
          contents: Buffer.from("#!/bin/sh\n"),
          executable: true,
          kind: "file",
          profile: "default",
          repoPath: "tool",
        }),
      ).resolves.toBe(true);
    },
  );

  it.skipIf(process.platform === "win32")(
    "reports executable artifact drift when the executable bit differs",
    async () => {
      const workspace = await createWorkspace();
      const artifactDirectory = join(workspace, "profiles", "default");
      const artifactPath = join(artifactDirectory, "tool");

      await mkdir(artifactDirectory, { recursive: true });
      await writeFile(artifactPath, "#!/bin/sh\n", "utf8");
      await chmod(artifactPath, 0o644);

      await expect(
        isRepoArtifactCurrent(workspace, {
          category: "plain",
          contents: Buffer.from("#!/bin/sh\n"),
          executable: true,
          kind: "file",
          profile: "default",
          repoPath: "tool",
        }),
      ).resolves.toBe(false);
    },
  );
});
